### ISSUE TITLE
-- lower precision computation floatX = float16, why not adding intX param in theano.config ?

### ISSUE THREAD

**Issue by jc-Rosichini** [Theano/Theano#5868]
Using a NVIDIA Pascal architecture provides instructions that operate on:
(1) 16-bit floating point data  addressed in theano by defining floatX = float16
(2),8- and 16-bit integer data (INT8 and INT16), would this be possible to select them by adding a intX = intxx ?
thks

**Comment by nouiz** [comment]
First, Theano currently only support float16 storage, not computation. It wouldn't be hard to just allow floatX computation for cuda lib like cudnn and cublas, but it isn't done yet.

For integer, I'm not sure we need a type that will vary automatically. Until we see a real need for a intX, you can just do it yourself like this:

intX = 'int8'

and use intX where you need to specify a dtype. Then if later Theano add this as a flag, you could just change that line to something like `from theano.config import intX`.

Why do you think it is important to add an intX as a Theano flag? Why that would be used frequently enough that it need support in Theano for it?

**Comment by jc-Rosichini** [comment]
Well defining my own intX ='int**' is what I'm doing currently.

The reason I'm shrinking as much as possible the data size is (a) to limit the bandwith exchanges with the GPU and (b) to take advantage of cudnn limited precision computation.

For "int**" since for instance I'm using a lot classes indices for sort of semantic segmentation this is a valuable approach.

btw.
(1) you mention that Theano currently only support float16 storage, not computation?, I may have wrongly understood that this was already built in the new git... and the appropriate kernel code generated.

(2) in this case, what should be understood with the following optimizer messages? Are there only related to storage support? I'm confused ;-)

`Compilation starts at 2017-04-19 16:32:36
optimizer: rmsprop, initial learning rate:  0.001
Disabling C code for Eye{dtype='float16'} due to unsupported float16
Disabling C code for Elemwise{true_div} due to unsupported float16
Disabling C code for Elemwise{mul} due to unsupported float16
Disabling C code for Argmax due to unsupported float16
Disabling C code for GpuSplit{2} due to unsupported float16
Disabling C code for GpuSplit{2} due to unsupported float16
...
Disabling C code for GpuSplit{2} due to unsupported float16
Disabling C code for GpuSplit{5} due to unsupported float16
...
Disabling C code for GpuSplit{15} due to unsupported float16
...
train compilation took 158.606 seconds
Compilation starts at 2017-04-19 16:35:16
Disabling C code for Eye{dtype='float16'} due to unsupported float16
Disabling C code for DeepCopyOp due to unsupported float16
Disabling C code for Argmax due to unsupported float16
valid compilation took 24.674 seconds`

Regards
JC

**Comment by nouiz** [comment]
To implement safely float16 without forgetting corner case that would give
wrong results, by defauls if one op have float16 inputs, we disable its c
code and fall back to the perform() method (python code, numpy seem to have
some computation in float16, so most of the time it should work, but
slowly). When this happen, it print a warning that the c code is disable
for that op.

When we update an op to support float16, we add to it an attribute that
tell it support float16 c codes. We did not add much CPU support for that.
So most CPU ops will have there c code disabled. But we did so for
many/most of the GPU ops. In your error, I found that GpuSplit don't have
support for it.

The way Theano added the support of float16 to the GPU ops, is that when we
read/write to the GPU memory, we convert to float32 and do the computation
in float32. This way, all GPUs can have the storage saving.

Sadly, it is harder to implement float16 computation, as the c language
don't define a float16 type. So instead of doing x+y, we need to call
functions like cuda_add_float16(x, y). A work around is to make a class
that implement + and other such operators. This isn't done.

Getting the speed up from CUBLAS and CUDNN is probably trivial, just pass
the right parameter to those functions or call the right function. But we
didn't do it yet. We aren't sure of the user interface for now.

But anyway, the first step is to make it work for float16 storage! Only
when that one is done moving to float16 computation make sence

warning, I have heard it isn't trivial to use float16 for computation. It
have not enough precission in many cases and need adjustment to the model.
In GTC, NVIDIA have a talk about that.

I'll check about GpuSplit, why we fall back to perform() method, but this
is just a missed speed up. it should work for that op.
